File: backend/pookie-backend/app/schemas/something.py
from pydantic import BaseModel, Field, ConfigDict, field_serializer
from datetime import datetime, timezone
from typing import Optional, List
from enum import Enum
from uuid import UUID


class ContentType(str, Enum):
    text = "text"
    image = "image"
    video = "video"
    url = "url"


class SomethingBase(BaseModel):
    model_config = ConfigDict(use_enum_values=True)

    content: Optional[str] = None
    content_type: ContentType = Field(default=ContentType.text, serialization_alias="contentType")
    media_url: Optional[str] = Field(default=None, serialization_alias="mediaUrl")


class SomethingCreate(SomethingBase):
    pass


class CirclePrediction(BaseModel):
    """Circle prediction from centroid similarity."""

    circle_id: int = Field(serialization_alias="circleId")
    circle_name: str = Field(serialization_alias="circleName")
    confidence: float


class SomethingResponse(SomethingBase):
    model_config = ConfigDict(
        from_attributes=True,
        use_enum_values=True
    )

    id: int
    user_id: UUID | str = Field(serialization_alias="userId")
    meaning: Optional[str] = None
    is_meaning_user_edited: bool = Field(default=False, serialization_alias="isMeaningUserEdited")
    novelty_score: Optional[float] = Field(default=None, ge=0.0, le=1.0, serialization_alias="noveltyScore")
    suggested_circles: List[CirclePrediction] = Field(default_factory=list, serialization_alias="suggestedCircles")
    created_at: datetime = Field(serialization_alias="createdAt")
    updated_at: datetime = Field(serialization_alias="updatedAt")

    @field_serializer('user_id')
    def serialize_user_id(self, user_id: UUID | str) -> str:
        """Convert UUID to string for JSON serialization."""
        return str(user_id)

    @field_serializer('created_at')
    def serialize_datetime(self, created_at: datetime) -> str:
        """Ensure datetime is always ISO8601 string for JSON output."""
        return created_at.isoformat(timespec="microseconds").replace("+00:00", "Z")

    
    @field_serializer('updated_at')
    def serialize_datetime(self, updated_at: datetime) -> str:
        """Ensure datetime is always ISO8601 string for JSON output."""
        return updated_at.isoformat(timespec="microseconds").replace("+00:00", "Z")

class SomethingUpdateMeaning(BaseModel):
    meaning: str

File: ios/Pookie/Pookie/Models/Something.swift
//
//  Something.swift
//  Pookie
//
//  Created by Dev Agent on 12/7/25.
//

import Foundation

// MARK: - Content Type Enum

/// Multimodal content type for somethings
///
/// Defines the type of content captured in a something. This enables the app to handle
/// text, images, videos, and URLs with appropriate processing and display logic.
public enum ContentType: String, Codable {
    /// Plain text content - user typed or voice-to-text transcription
    /// Example: "Remember to buy milk"
    case text

    /// Image content - photo captured or selected from library
    /// Requires: mediaUrl pointing to image file
    case image

    /// Video content - recorded or selected from library
    /// Requires: mediaUrl pointing to video file
    case video

    /// URL content - web link or bookmark
    /// Content field contains the URL, mediaUrl may contain preview image
    case url
}

// MARK: - Circle Prediction

/// Circle prediction from centroid similarity
public struct CirclePrediction: Codable, Equatable {
    public let circleId: Int
    public let circleName: String
    public let confidence: Double
}

// MARK: - Something Model

/// Core Something model matching backend API schema
public struct Something: Codable, Identifiable, Equatable {
    public let id: Int
    public let userId: String
    public let content: String?
    public let contentType: ContentType
    public let mediaUrl: String?
    public let meaning: String?
    public let isMeaningUserEdited: Bool
    public let noveltyScore: Double?
    public let suggestedCircles: [CirclePrediction]
    public let createdAt: Date
    public let updatedAt: Date

}

// MARK: - Request/Response Models

/// Request body for creating a new something
public struct SomethingCreate: Codable {
    public let content: String?
    public let contentType: ContentType
    public let mediaUrl: String?
    
    public init(content: String?, contentType: ContentType, mediaUrl: String?) {
        self.content = content
        self.contentType = contentType
        self.mediaUrl = mediaUrl
    }
}

/// Request body for updating something meaning
public struct SomethingUpdateMeaning: Codable {
    public let meaning: String

    public init(meaning: String) {
        self.meaning = meaning
    }
}
